HeterogeneousCore/SonicTriton/src/TritonService.cc

```diff
@@ -0,0 +1,293 @@
+#include "HeterogeneousCore/SonicTriton/interface/TritonService.h"
+#include "HeterogeneousCore/SonicTriton/interface/triton_utils.h"
+
+#include "DataFormats/Provenance/interface/ModuleDescription.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ServiceRegistry/interface/ActivityRegistry.h"
+#include "FWCore/ServiceRegistry/interface/ProcessContext.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include "grpc_client.h"
+#include "grpc_service.pb.h"
+
+#include <cstdio>
+#include <cstdlib>
+#include <filesystem>
+#include <utility>
+#include <tuple>
+
+namespace ni = nvidia::inferenceserver;
+namespace nic = ni::client;
+
+const std::string TritonService::Server::fallbackName{"fallback"};
+const std::string TritonService::Server::fallbackUrl{"0.0.0.0:8001"};
+
+namespace {
+  std::pair<std::string, int> execSys(const std::string& cmd) {
+    //redirect stderr to stdout
+    auto pipe = popen((cmd + " 2>&1").c_str(), "r");
+    if (!pipe)
+      throw cms::Exception("SystemError") << "popen() failed for command: " << cmd;
```

Comment:
The documentation says `errno` may be set if the open fails. Would be good to get that information.